compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+
+using namespace mlir;
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+namespace {
+
+struct FuseHorizontalContractionsPass
+    : public FuseHorizontalContractionsBase<FuseHorizontalContractionsPass> {
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, IREE::Flow::FlowDialect,
+                    math::MathDialect>();
+  }
+  FuseHorizontalContractionsPass() {}
+  FuseHorizontalContractionsPass(const FuseHorizontalContractionsPass &pass)
+      : FuseHorizontalContractionsPass() {}
+
+  void runOnOperation() override;
+};
+
+} // namespace
+
+/// Structs that captures the ops that are to be fused
+struct HorizontalFusionGroup {
+  // Contractions op that are to be fused.
+  SmallVector<linalg::LinalgOp> contractionOps;
+  // Optional truncate operations that could be following the contraction op.
+  std::optional<SmallVector<linalg::GenericOp>> truncateOps;
+  // Operation that dominates all the ops of the group.
+  Operation *dominatedByAll;
+};
+
+/// Get user of operation that is a truncate operation.
+static std::optional<linalg::GenericOp> getTruncUser(Operation *op) {
+  if (op->getNumResults() != 1) {
+    return std::nullopt;
+  }
+  Value result = op->getResult(0);
+  if (!result.hasOneUse()) {
+    return std::nullopt;
+  }
+  Operation *user = *result.user_begin();
+  auto genericOp = dyn_cast<linalg::GenericOp>(user);
+  if (!genericOp) {
+    return std::nullopt;
+  }
+  if (genericOp.getNumDpsInputs() != 1 || genericOp.getNumDpsInits() != 1) {
+    return std::nullopt;
+  }
+  if (llvm::any_of(genericOp.getIndexingMapsArray(),
+                   [](AffineMap map) { return !map.isIdentity(); })) {
+    return std::nullopt;
+  }
+  if (genericOp.getNumParallelLoops() != genericOp.getNumLoops()) {
+    return std::nullopt;
+  }
+  auto yieldOp = cast<linalg::YieldOp>(genericOp.getBody()->getTerminator());
+  auto yieldDefOp = yieldOp->getOperand(0).getDefiningOp<arith::TruncFOp>();
+  if (!yieldDefOp) {
+    return std::nullopt;
+  }
+  auto arg = dyn_cast<BlockArgument>(yieldDefOp->getOperand(0));
+  if (!arg || arg.getParentBlock() != genericOp.getBody() ||
+      arg.getArgNumber() != 0) {
+    return std::nullopt;
+  }
+  return genericOp;
+}
+
+/// Find the operation that is dominated by all other operation.
+std::optional<Operation *>
+findDominatedByAllOps(const SetVector<Operation *> &ops,
+                      const DominanceInfo &dominanceInfo) {
+  for (auto opA : ops) {
+    bool opAIsDominatedByAll = true;
+    for (auto opB : ops) {
+      if (opA == opB) {
+        continue;
+      }
+      if (!dominanceInfo.properlyDominates(opB, opA)) {
+        opAIsDominatedByAll = false;
+        break;
+      }
+    }
+    if (opAIsDominatedByAll) {
+      return opA;
+    }
+  }
+  return std::nullopt;
+}
+
+/// Find all candidates that can be used for horizontal fusion
+static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
```

Comment:
I think it is worths to document what the "candidates" means, so others (like me) don't need to look at all the implementation details.